File: scripts/build_video_manifest.mjs
import { createHash } from "node:crypto";
import { constants as fsConstants } from "node:fs";
import { copyFile, link, readdir, stat, mkdir, writeFile } from "node:fs/promises";
import path from "node:path";

const VIDEO_EXTENSIONS = new Set([".mp4", ".mov", ".avi", ".mkv", ".webm", ".m4v"]);
const TIME_PATTERN = /__(\d+(?:\.\d+)?)s__(\d+(?:\.\d+)?)s$/i;

function stableId(relativePath) {
  const digest = createHash("sha1").update(relativePath, "utf8").digest("hex").slice(0, 12);
  return `chronophy_video_${digest}`;
}

function parseName(filePath) {
  const parsed = path.parse(filePath);
  let title = parsed.name;
  let clipStartSec = null;
  let clipEndSec = null;
  let clipDurationSec = null;

  const timeMatch = title.match(TIME_PATTERN);
  if (timeMatch) {
    clipStartSec = Number(timeMatch[1]);
    clipEndSec = Number(timeMatch[2]);
    clipDurationSec = Number((clipEndSec - clipStartSec).toFixed(3));
    title = title.slice(0, timeMatch.index);
  }

  let sourceVideoId = null;
  if (title.includes("#")) {
    const parts = title.split("#");
    sourceVideoId = parts.shift() || null;
    title = parts.join("#");
  } else {
    const prefix = title.match(/^[A-Za-z0-9_-]{8,}/);
    if (prefix) {
      sourceVideoId = prefix[0];
      title = title.slice(prefix[0].length);
    }
  }

  return {
    source_video_id: sourceVideoId,
    title: title.trim(),
    clip_start_sec: clipStartSec,
    clip_end_sec: clipEndSec,
    clip_duration_sec: clipDurationSec,
  };
}

async function walk(root) {
  const out = [];
  async function visit(current) {
    const entries = await readdir(current, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = path.join(current, entry.name);
      if (entry.isDirectory()) {
        await visit(fullPath);
      } else if (entry.isFile() && VIDEO_EXTENSIONS.has(path.extname(entry.name).toLowerCase())) {
        out.push(fullPath);
      }
    }
  }
  await visit(root);
  return out.sort((a, b) => a.localeCompare(b, "zh-Hans-CN"));
}

function parseArgs(argv) {
  const args = {
    root: path.resolve("..", "剪辑后视频"),
    base: path.resolve(".."),
    output: path.resolve("data", "video_manifest.jsonl"),
    privateMapping: path.resolve("data", "private_video_mapping.jsonl"),
    releaseDir: null,
    limit: null,
    copyMode: "none",
  };
  for (let i = 0; i < argv.length; i += 1) {
    const key = argv[i];
    const value = argv[i + 1];
    if (key === "--root") args.root = path.resolve(value);
    if (key === "--base") args.base = path.resolve(value);
    if (key === "--output") args.output = path.resolve(value);
    if (key === "--private-mapping") args.privateMapping = path.resolve(value);
    if (key === "--release-dir") args.releaseDir = path.resolve(value);
    if (key === "--limit") args.limit = Number(value);
    if (key === "--copy-mode") args.copyMode = value;
    if (key.startsWith("--")) i += 1;
  }
  return args;
}

const args = parseArgs(process.argv.slice(2));
const files = await walk(args.root);
const selected = args.limit ? files.slice(0, args.limit) : files;
const rows = [];
const privateRows = [];

for (const [index, filePath] of selected.entries()) {
  const info = await stat(filePath);
  const relativePath = path.relative(args.base, filePath).replaceAll(path.sep, "/");
  const relativeToRoot = path.relative(args.root, filePath).split(path.sep);
  const videoId = `chronophy_${String(index + 1).padStart(6, "0")}`;
  const publicFileName = `${videoId}${path.extname(filePath).toLowerCase()}`;
  const publicPath = `videos/${publicFileName}`;
  const parsed = parseName(filePath);

  rows.push({
    id: videoId,
    video_id: videoId,
    path: publicPath,
    file_name: publicFileName,
    file_size_bytes: info.size,
    clip_start_sec: parsed.clip_start_sec,
    clip_end_sec: parsed.clip_end_sec,
    clip_duration_sec: parsed.clip_duration_sec,
    split: "unassigned",
  });

  privateRows.push({
    id: videoId,
    original_stable_id: stableId(relativePath),
    public_path: publicPath,
    original_path: relativePath,
    original_source_collection: relativeToRoot[0] || "",
    original_file_name: path.basename(filePath),
    original_source_video_id: parsed.source_video_id,
    original_title: parsed.title,
    file_size_bytes: info.size,
  });

  if (args.releaseDir) {
    const target = path.join(args.releaseDir, publicFileName);
    await mkdir(path.dirname(target), { recursive: true });
    if (args.copyMode === "hardlink") {
      try {
        await link(filePath, target);
      } catch (error) {
        if (error.code !== "EEXIST") throw error;
      }
    } else if (args.copyMode === "copy") {
      await copyFile(filePath, target, fsConstants.COPYFILE_EXCL).catch((error) => {
        if (error.code !== "EEXIST") throw error;
      });
    }
  }
}

await mkdir(path.dirname(args.output), { recursive: true });
await writeFile(args.output, rows.map((row) => JSON.stringify(row)).join("\n") + "\n", "utf8");
await mkdir(path.dirname(args.privateMapping), { recursive: true });
await writeFile(
  args.privateMapping,
  privateRows.map((row) => JSON.stringify(row)).join("\n") + "\n",
  "utf8",
);
console.log(`Wrote ${rows.length} public rows to ${args.output}`);
console.log(`Wrote ${privateRows.length} private mapping rows to ${args.privateMapping}`);
if (args.releaseDir) {
  console.log(`${args.copyMode} release files in ${args.releaseDir}`);
}

File: scripts/build_video_manifest.py
"""Build a JSONL manifest from a directory of clipped videos.

The script is intentionally conservative: it does not copy videos, rename files,
or modify the source directory. It only records stable metadata that can be used
by later annotation and release steps.
"""

from __future__ import annotations

import argparse
import hashlib
import json
import re
from pathlib import Path
from typing import Any


VIDEO_EXTENSIONS = {".mp4", ".mov", ".avi", ".mkv", ".webm", ".m4v"}
TIME_PATTERN = re.compile(r"__(?P<start>\d+(?:\.\d+)?)s__(?P<end>\d+(?:\.\d+)?)s$", re.I)


def stable_id(relative_path: str) -> str:
    digest = hashlib.sha1(relative_path.encode("utf-8")).hexdigest()[:12]
    return f"chronophy_video_{digest}"


def parse_name(path: Path) -> dict[str, Any]:
    stem = path.stem
    start = None
    end = None
    title = stem

    match = TIME_PATTERN.search(stem)
    if match:
        start = float(match.group("start"))
        end = float(match.group("end"))
        title = stem[: match.start()]

    source_video_id = None
    if "#" in title:
        source_video_id, title = title.split("#", 1)
    else:
        prefix = re.match(r"^[A-Za-z0-9_-]{8,}", title)
        if prefix:
            source_video_id = prefix.group(0)
            title = title[prefix.end() :]

    return {
        "source_video_id": source_video_id,
        "title": title.strip(),
        "clip_start_sec": start,
        "clip_end_sec": end,
        "clip_duration_sec": round(end - start, 3) if start is not None and end is not None else None,
    }


def iter_videos(root: Path) -> list[Path]:
    return sorted(
        path
        for path in root.rglob("*")
        if path.is_file() and path.suffix.lower() in VIDEO_EXTENSIONS
    )


def build_manifest(root: Path, base: Path) -> list[dict[str, Any]]:
    rows: list[dict[str, Any]] = []
    for path in iter_videos(root):
        relative_path = path.relative_to(base).as_posix()
        source_parts = path.relative_to(root).parts
        source_collection = source_parts[0] if source_parts else ""
        parsed = parse_name(path)
        row = {
            "id": stable_id(relative_path),
            "path": relative_path,
            "source_collection": source_collection,
            "file_name": path.name,
            "file_size_bytes": path.stat().st_size,
            **parsed,
        }
        rows.append(row)
    return rows


def main() -> None:
    parser = argparse.ArgumentParser(description="Build ChronoPhy video manifest.")
    parser.add_argument("--root", type=Path, default=Path("../剪辑后视频"), help="Video dataset root.")
    parser.add_argument("--base", type=Path, default=Path(".."), help="Base directory for relative paths.")
    parser.add_argument("--output", type=Path, default=Path("data/video_manifest.jsonl"))
    parser.add_argument("--limit", type=int, default=None, help="Optional maximum number of rows.")
    args = parser.parse_args()

    root = args.root.resolve()
    base = args.base.resolve()
    rows = build_manifest(root, base)
    if args.limit is not None:
        rows = rows[: args.limit]

    args.output.parent.mkdir(parents=True, exist_ok=True)
    with args.output.open("w", encoding="utf-8") as handle:
        for row in rows:
            handle.write(json.dumps(row, ensure_ascii=False) + "\n")

    print(f"Wrote {len(rows)} rows to {args.output}")


if __name__ == "__main__":
    main()
